[Docs] Combine All @elastic/eui Imports in the EUI Docs Demo JS Tab

breehall opened the pull request:
### Summary
Fixes #5104 . Previously, when multiple @elastic/eui imports were included in a module, the function to prepare and render the code for the Demo JS tab within the EUI docs would not combine some, but not all of the imports. It seems that there was logic missing to capture and combine imports that spanned multiple lines. I've organized the `renderJsSourceCode` method in `_utils.js` to combine all @elastic/eui imports and improve the readability of the function over all. 

Below are images of the imports on the Forms and Form Rows page before and after the logic changes. 

Before
![A7A2568C-BE98-4759-A9DD-629188D05FE6](https://user-images.githubusercontent.com/40739624/149217947-192ed00c-dd24-4294-acc5-f069b4ed05ef.jpeg)

After
![E5489FAD-B6A5-432D-80E9-2797F1905774](https://user-images.githubusercontent.com/40739624/149217987-647e4c53-55cc-45b3-9e9e-8b964b7f2d1c.jpeg)

### Checklist
~[ ] Check against **all themes** for compatibility in both light and dark modes~
~[ ] Checked in **mobile**~
~[ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**~
- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**

~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
~[ ] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**~

- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
~[ ] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately~

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

thompsongl reviewed (commented):
Cool! Looks like the consolidation works, but the formatting might cause lint errors in codesandbox.
It's a line length setting (I think) and I suspect that codesandbox will show red lines when each import is not on its own line.
This could be tricky to test because codesandbox will be outdated until these changes have been released 🙃

Perhaps I'm overthinking this so I'll let others chime in, also.

breehall commented:
> Cool! Looks like the consolidation works, but the formatting might cause lint errors in codesandbox. 

Ah ok! I can update the consolidation to add a new line after each import (except the last) if that would give us better odds at avoiding issues with CodeSandbox.

constancecchen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -1,38 +1,29 @@
+ * EUI Docs. In addition to formatting the code for the tab, this function also combines EUI imports by
+ * searching code.default for all EUI imports, extracting the variables, and combining them at the top of
+ * the formatted code.
+ */
```
Appreciate these extra documentation notes, it makes it easy for someone new like me coming in to understand the intention & goals!

constancecchen reviewed (commented):
++ to newlines between each import - even in the Demo JS tab and not just CodeSandbox, it looks visually more pleasant/easier to read

constancecchen reviewed (changes requested):
Whoops, looks like there's an issue with imports from 3rd party libs - they're not getting a closing `}` bracket or new lines.

https://eui.elastic.co/pr_5533/#/tabular-content/data-grid

<img width="1087" alt="" src="https://user-images.githubusercontent.com/549407/149553435-4c3615b9-7de2-4eb7-92da-da88b7c2a6a0.png">

chandlerprall commented:
Good catch @constancecchen ! I've pushed up a fix to avoid processing non-eui imports. I expected the non-greedy dot-all match (`\{(.+?)\}`) to suffice, but with the left anchor being `import` and the right being `from '@elastic/eui'`, this setup:

```
import { thing } from 'somewhere';
import { EuiButton } from '@elastic/eui';
```

would match against the first import and consume all the way through the end of elastic/eui. I updated the regex to match against everything that isn't a `}` for better anchoring, and threw in a couple other clean ups that change opened up. I also documented what each bit of the regex is doing.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

constancecchen commented:
Love the new regex comments!!!

Sorry to be the worst, I'm just randomly clicking around to random demo js tabs and it looks like relative imports need an extra new line:

<img width="1034" alt="" src="https://user-images.githubusercontent.com/549407/149588159-ed233d1f-31ec-40b9-8fde-08f7ea1a7a87.png">

This one is from https://eui.elastic.co/pr_5533/#/tabular-content/tables

I'd also still love to see newlines between each `,` import item in `@elastic/eui` personally!

breehall commented:
> I'd also still love to see newlines between each `,` import item in `@elastic/eui` personally!

I'm working on this as I type this. Just doing some testing to see if it works!

breehall commented:
I just pushed a change to move each Elastic import to its own line. This question will likely boil down to personal preference, but I noticed when I add additional space after the EUI imports, the other imports don't have that same space. This made the imports section look inconsistent to me, so I removed the same. Now all the imports are on new lines with no additional space between the lines. Does anyone have a strong opinion about this?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

constancecchen commented:
> Now all the imports are on new lines with no additional space between the lines.

I think I see this now, I'm fine with it personally - in general I don't mind all imports being together without 2 newlines between each.

There's still a couple (separate, I think?) issues that I can see that need to be addressed:

<img width="610" alt="" src="https://user-images.githubusercontent.com/549407/149987964-e678dbe9-42a3-4737-9587-c2dfa72cb907.png">

Possibly the 2nd highlighted area (relative imports need a newline) is actually that the regex for absolute imports need a newline afterwards instead, not totally sure 🤷‍♀️

chandlerprall commented:
> Possibly the 2nd highlighted area (relative imports need a newline) is actually that the regex for absolute imports need a newline afterwards instead, not totally sure 🤷‍♀️

This was my fault, by adding the `[\r\n]` at the end of the regex it became too aggressive. Should be fine to remove.

breehall commented:
How does this look? This example is from Data grid schemas and popovers (http://localhost:9000/#/tabular-content/data-grid-schemas-and-popovers). 

![2DE101D7-913A-4271-B0B0-1764C887D933_4_5005_c](https://user-images.githubusercontent.com/40739624/150021582-7575a11f-2b19-4283-b18f-13172dbd4b78.jpeg)

@constancecchen which example did you use? I'd like to cross check the changes locally before pushing them.

constancecchen commented:
I was just checking the basic Data Grid and Tables docs. The staging links are in previous comments above!

Screenshot looks great for newlines! As a heads up though unless my eyesight is going I think the indentation on the EUI imports have 3 spaces instead of 2 🤔

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

constancecchen commented:
I'm not sure why or how, but it looks like this PR is adding an extra newline/whitespace at the very bottom of the demo JS code that doesn't exist on prod:

https://eui.elastic.co/pr_5533/#/tabular-content/tables:

<img width="370" alt="" src="https://user-images.githubusercontent.com/549407/150176050-d70d58de-8a81-484c-90da-2e037bddb1f0.png">

https://elastic.github.io/eui/#/tabular-content/tables:

<img width="388" alt="" src="https://user-images.githubusercontent.com/549407/150176066-6e18470b-622a-4069-a1ac-76c01ecda117.png">

breehall commented:
> I'm not sure why or how, but it looks like this PR is adding an extra newline/whitespace at the very bottom of the demo JS code that doesn't exist on prod:

It looks like this was due to an extra space and new line in the template string responsible for displaying the rendered code. I've updated it and checked it locally and will do the same once the PR preview is complete.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

breehall commented:
I just did a spot check with the Tables and Data Grid examples and I think it looks good now.

constancecchen commented:
Wohoo, can confirm the extra newline was fixed!

I hate to do this, but it looks like there are some double newlines occuring on a few demos:

<img width="551" alt="" src="https://user-images.githubusercontent.com/549407/150192522-7020ff11-809a-4167-8236-cb06ad99020c.png">

Examples: https://eui.elastic.co/pr_5533/#/display/title, https://eui.elastic.co/pr_5533/#/display/stat

It looks like the newline is coming from source code that looks like this:

```
import React from 'react';

import {
  EuiTitle,
  EuiHorizontalRule,
  EuiSpacer,
  EuiCode,
} from '../../../../src/components';
```

So basically the preceding newline between EUI imports and non-EUI aren't getting trimmed and are causing an extra newline to exist between the imports and the actual component. It's not the worst thing in the world so it's not a blocker - I'll leave that up to y'all to decide if you'd rather just move ahead or attempt to fix it now in the regex.

breehall commented:
> So basically the preceding newline between EUI imports and non-EUI aren't getting trimmed and are causing an extra newline to exist between the imports and the actual component. It's not the worst thing in the world so it's not a blocker - I'll leave that up to y'all to decide if you'd rather just move ahead or attempt to fix it now in the regex.

I'm going to be tackling this in an additional PR that will add another regex or function to format the non-Eui imports.

constancecchen reviewed (approved):
Sounds great, thanks for your patience y'all!

thompsongl commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -1,38 +1,39 @@
+
+  // Find all imports that come from '@elastic/eui'
+  renderedCode = renderedCode.replace(
+    // [\r\n]                       - start of a line
+    // import\s+\{                  - import / whitespace / opening brace
+    // ([^}]+)                      - group together anything that isn't a closing brace
+    // \}\s+from\s+'@elastic\/eui'; - closing brace / whitespace / from / whitespace / '@elastic/eui';
+    // [\r\n]                       - match end of line, so the extra new line is removed via the replace operation
```
💟

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

breehall commented:
Hey @constancecchen @thompsongl @chandlerprall - do you all mind re-reviewing this PR? I ended up making changes ti the existing functionality as well as creating new functionality for this PR. 

**Here's a quick run down:** 

1. I updated the existing Eui import formatting to take import statements that bring in only one component/function/utility into a file. In the original version of this PR, since imports were pushed onto a new line.
![6522816F-D24A-420C-B0B7-F4B9E4EBA8CB](https://user-images.githubusercontent.com/40739624/151640702-fae43b6a-55bf-49bc-b9d0-fd213b0718fc.jpeg)

2. I added functionality to give the non-Eui imports the luxury formatting treatment! This means all imports should be formatted uniformly. This was a little tricky because there were four formats the the import statements could appear in. Below are examples of the patterns I found. From what I can tell, each pattern has been accounted for in formatting,
A. `import React from 'react';`
B. `import { fake } from 'faker';`
C. `import React, { Fragment, useState } from 'react';`
D. `import { test1, test2 } from 'test-library';`
![E654A75E-199F-48C5-8ED1-FF5D01D28F90](https://user-images.githubusercontent.com/40739624/151641076-5d6c939e-6890-447e-a3a7-f74491957fd8.jpeg)

3. I normalized the spacing between the import statements and the start of the demo code. In the first version of this PR, there was some additional white space added between the imports and the start of the code blocks. There's now a single new line separating the imports and the code.

Thank you all!

constancecchen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -1,38 +1,92 @@
+  elasticImports = elasticImports.filter((ele) => ele);
+
+  let formattedEuiImports = '';
+  if (elasticImports.length === 1) {
+    // Account for and format Eui import statements that only contain one function/utility
+    formattedEuiImports = `import { ${elasticImports} } from '@elastic/eui';`;
+  } else {
```
Rather than automatically creating newlines for each import based on the # of import, would be it be possible to check for character/line length?

It looks like both CodeSandbox and our local demo JS's switch to newlines for each import once the character length of the line goes above 81 characters (including the semicolon). Would it be possible to create a DRY helper that checks both EUI and non-EUI imports for char length > 81 instead?

No worries if not or if that's too complicated!

breehall replied:
Sure, I can work on this. Do you think this will take away from the uniformity of the imports? The reason I added the logic to combine and format the non-EUI imports is because we had some code examples where imports were separated with a new line, and examples where the lines were short enough so they would just be on the same line. 

If this is something that we want to adjust, would it make more sense to just remove the non-EUI import logic and allow Prettier to handle the formatting when the example is being developed?

For example, let's say we have the two import statements below:

```
import React, { useState } from 'react';
```

```
import { EuiFlyout, EuiFlyoutBody, EuiFlyoutHeader, EuiButton, EuiText, EuiTitle, EuiCodeBlock, } from '@elastic/eui';
```

If a code demo was created with these two import statements, when `yarn lint-fix` is run before the dev commits their code, the first import would remain the same and the second would be broken up with new lines. Please correct me if I'm wrong, but it feel like checking the length of the import statement before running the logic to separate imports with new lines is duplicating work that the linter would do for us.

This would have to be done to the EUI imports because they don't get linted, but should we be doing it for the non-EUI imports?

constancecchen replied:
> Do you think this will take away from the uniformity of the imports?

No, I don't think this is the same thing as the fact we have demo JS with multiple lines/imports for EUI imports. This is a Prettier/char length rule thing just IMO.

> Please correct me if I'm wrong, but it feel like checking the length of the import statement before running the logic to separate imports with new lines is duplicating work that the linter would do for us.

Normally yes, but I believe this needs to be manually checked for EUI imports because unfortunately we have a bunch of demo JS doing shenanigans like this:

```
import { EuiFlyout } from '../../../src/components/flyout';
import { EuiCode } from '../../../src/components/code';
import { EuiText, EuiButton } from '../../../src/components';
```

Those lines should all get converted to `@elastic/eui` and combined+line checked, there isn't a realistic way a linter could handle this because it doesn't know that's what we want to do.

> If this is something that we want to adjust, would it make more sense to just remove the non-EUI import logic and allow Prettier to handle the formatting when the example is being developed?

Sure, if we want! I'm not actually 100% sure what the non-EUI import logic is doing 🙈 It's nice to have a regex rule for non-EUI imports make newlines more consistent between each import statement but I agree line char-length can be left up to Prettier.

breehall replied:
> Those lines should all get converted to @elastic/eui and combined+line checked, there isn't a realistic way a linter could handle this because it doesn't know that's what we want to do.

I added a helper that is being used to check the length of the import statements. When the statement is under 81 characters, the EUI imports are composed on one line. Otherwise, the imports use the existing logic to break into new lines.

> Sure, if we want! I'm not actually 100% sure what the non-EUI import logic is doing 🙈 It's nice to have a regex rule for non-EUI imports to make newlines more consistent between each import statement but I agree line char-length can be left up to Prettier.

I also removed the non-EUI formatting logic, so we're just running that helper on the EUI imports and we will leave the non-EUI formatting up to Prettier.

constancecchen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -1,38 +1,92 @@
-      renderedCode[idx] = '';
+  let renderedCode = cleanEuiImports(code.default);
+
+  /* ----- Combine and clean Eui imports ----- */
```
super minor/optional note, I believe EUI should be capitalized as an initialism

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

constancecchen commented:
Awesome, this is looking fantastic!! One comment from a quick QA - it looks like newlines between imports aren't being preserved as expected, leading to double newlines between source and final JS. Example in `src-docs/src/views/date_picker/date_picker.js`:

Input:
```
import React, { useState } from 'react';

import moment from 'moment';

import { EuiDatePicker, EuiFormRow } from '../../../../src/components';

export default () => {
```

Output:

```
import { EuiDatePicker, EuiFormRow } from '@elastic/eui';
import React, { useState } from 'react';

import moment from 'moment';


export default () => {
```

Is there any way to either reasonably preserve newlines between imports, or at worst trim all of them instead of just some?

breehall commented:
> Is there any way to either reasonably preserve newlines between imports, or at worst trim all of them instead of just some?

Sure, this is something that was included in the non-EUI import formatting logic that I had before this version of the PR. It standardized the import styling, but also removed extra spaces so they could be uniformly added when composing the EUI imports, non-EUI imports, and remaining code. Should I add this piece back in?

constancecchen commented:
If it's possible to add that logic back in but just to handle newlines, that would be great!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/

breehall commented:
Yep! What I did this time is include the regex for the non-EUI imports, but the only thing I did was combine them with a single new line in between statements to for standardization. There is no formatting of these imports or checks for line length for non-EUI imports. 

Example from `src-docs/src/views/date_picker/date_picker.js`.


![C15ED608-195A-4EA0-BCFC-6D6116A82549_4_5005_c](https://user-images.githubusercontent.com/40739624/152570123-9c367d92-28ea-414f-bbc0-611dfaa94a97.jpeg)

constancecchen commented on src-docs/src/components/guide_section/_utils.js:
```diff
@@ -1,38 +1,73 @@
+  let formattedEuiImports = '';
+
+  // determine if imports should be wrapped to new lines based on the import statement length
+  if (wrapImportsOnNewLines(elasticImports)) {
```
Optional refactor: I definitely appreciate this being a separate check, but it feels like there's a lot of repeated code between this method and lines 42-43 which would break if we updated the syntax in one place but not another. Maybe consider branching logic like this instead:

```js
const combinedImports = elasticImports.join(', ');
const singleLineImports = `import { ${combinedImports} } from '@elastic/eui';`;

if (singleLineImports.length <= 81) {
  formattedEuiImports = singleLineImports;
} else {
  const lineSeparatedImports = elasticImports.join(',\n  ');
  formattedEuiImports = `import {\n  ${lineSeparatedImports},\n} from '@elastic/eui';`;
}
```

breehall replied:
That makes sense! Especially since the helper is only used in one place now. I'll make this change.

constancecchen replied:
Latest diff looks great!

breehall replied:
🙌🏾. Yay! I'll get this merged. Thank you!

constancecchen reviewed (approved):
QA looks fantastic! 🤩 Thanks so much for your patience with my back and forth on this Bree!!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5533/
